### PR TITLE
Add sticky parameter to the posts endpoint.

### PR DESCRIPTION
**Pull request by joehoyle**
This allows getting posts that are either "only sticky" for "not sticky"
via `sticky=true` or `sticky=false`. This can also be used in conjunction
with `include` and `exclude`.

### CODE REVIEW

**Comment by joehoyle** [comment]
Fixes #2705

File: tests/test-rest-posts-controller.php [hunk 1]
```diff
@@ -1441,6 +1520,19 @@ public function test_update_post_empty_content() {
...
+
+		$request = new WP_REST_Request( 'PUT', sprintf( '/wp/v2/posts/%d', $this->post_id ) );
+		$request->set_body_params( array(
+			'content' => 'Hello\\s',
+		) );
+
+		$response = $this->server->dispatch( $request );
+		$new_data = $response->get_data();
+		$this->assertEquals( 'Hello\\s', $new_data['content']['raw'] );
+	}
```
Comment (rmccue) [inline tests/test-rest-posts-controller.php]: This test appears to be unrelated to the rest of the PR, can you split it into a new one? :)

**Review by rmccue — CHANGES_REQUESTED** [review]

File: tests/test-rest-posts-controller.php (hunk 1 above)
Comment (joehoyle) [reply tests/test-rest-posts-controller.php]: Ohhhhhh that's why this is failing, doh! Didn't mean to commit this.

File: lib/endpoints/class-wp-rest-posts-controller.php [hunk 2]
```diff
@@ -1776,6 +1798,14 @@ public function get_collection_params() {
 				'default'           => array(),
 			);
 		}
+
+		if ( 'post' === $this->post_type ) {
+			$params['sticky'] = array(
+				'description'       => __( 'Limit result set to items that are sticky.' ),
+				'type'              => 'boolean',
```
Comment (rmccue) [inline lib/endpoints/class-wp-rest-posts-controller.php:1805]: Indentation is off.

**Review by rmccue — CHANGES_REQUESTED** [review]

**Comment by rmccue** [comment]
Over to KAdam for review.

**Review by rmccue — APPROVED** [review]

File: lib/endpoints/class-wp-rest-posts-controller.php [hunk 3]
```diff
@@ -124,6 +124,28 @@ public function get_items( $request ) {
 			unset( $args['filter'] );
 		}
 
+		if ( isset( $request['sticky'] ) ) {
```
Comment (kadamwhite) [inline lib/endpoints/class-wp-rest-posts-controller.php:127]: As noted in slack, `sticky=false` does not work (it's interpreted as `true` due to string coercion -- suggest using `rest_is_boolean` introduced in #2630

**Review by kadamwhite — CHANGES_REQUESTED** [review]

File: tests/test-rest-posts-controller.php [hunk 4]
```diff
@@ -357,6 +358,88 @@ public function test_get_items_tags_and_categories_query() {
...
+		$request = new WP_REST_Request( 'GET', '/wp/v2/posts' );
+		$request->set_param( 'sticky', false );
+
+		$response = $this->server->dispatch( $request );
+		$this->assertCount( 1, $response->get_data() );
+
+		$posts = $response->get_data();
+		$post = $posts[0];
+		$this->assertEquals( $id1, $post['id'] );
+	}
```
Comment (kadamwhite) [inline tests/test-rest-posts-controller.php:422]: Not sure if it's worth it but I think this test (variant of the one above) will demonstrate whether `sticky=false` in query string will work:

``` php
    public function test_get_items_not_sticky_query_string_boolean() {
        $id1 = $this->post_id;
        $id2 = $this->factory->post->create( array( 'post_status' => 'publish' ) );

        update_option( 'sticky_posts', array( $id2 ) );

        $request = new WP_REST_Request( 'GET', '/wp/v2/posts' );
        $request->set_param( 'sticky', 'false' );

        $response = $this->server->dispatch( $request );
        $this->assertCount( 1, $response->get_data() );

        $posts = $response->get_data();
        $post = $posts[0];
        $this->assertEquals( $id1, $post['id'] );
    }
```

File: lib/endpoints/class-wp-rest-posts-controller.php [hunk 5]
```diff
@@ -1776,6 +1798,14 @@ public function get_collection_params() {
 				'default'           => array(),
 			);
 		}
+
+		if ( 'post' === $this->post_type ) {
+			$params['sticky'] = array(
+				'description' => __( 'Limit result set to items that are sticky.' ),
+				'type'        => 'boolean',
```
Comment (BE-Webdesign) [inline lib/endpoints/class-wp-rest-posts-controller.php]: Add in the `rest_validate_request_arg` and `rest_sanitize_request_arg` for validate and sanitize callbacks and this patch will be da hotness!

**Review by BE-Webdesign — CHANGES_REQUESTED** [review]

File: tests/test-rest-posts-controller.php (hunk 4 above)
Comment (kadamwhite) [reply tests/test-rest-posts-controller.php:422]: After discussion w/ @BE-Webdesign & @rmccue testing this case independently is probably overkill once the right sanitize callback is put in place

File: lib/endpoints/class-wp-rest-posts-controller.php (hunk 5 above)
Comment (joehoyle) [reply lib/endpoints/class-wp-rest-posts-controller.php]: This will automajically be done based off the schema already.

Comment (joehoyle) [reply lib/endpoints/class-wp-rest-posts-controller.php]: Actually this is not the schema, sorry my bad!

**Comment by joehoyle** [comment]
@rmccue @BE-Webdesign @kadamwhite ok I pushed an update to fix the `false` in the URL thang.

**Review by kadamwhite — APPROVED** [review]

**Comment by kadamwhite** [comment]
+1 from me, `sticky=false` now works as expected

**Review by BE-Webdesign — APPROVED** [review]

**Pull request merged**